clang/lib/Sema/SemaChecking.cpp

```diff
@@ -5573,6 +5576,122 @@ bool Sema::BuiltinSetjmp(CallExpr *TheCall) {
   return false;
 }
 
+ExprResult Sema::BuiltinCountedByRef(ExprResult TheCallResult) {
+  CallExpr *TheCall = cast<CallExpr>(TheCallResult.get());
+
+  // For simplicity, we support only a limited expressions for the argument.
+  // Specifically 'ptr->array' and '&ptr->array[0]'. This allows us to reject
+  // arguments with complex casting, which really shouldn't be a huge problem.
+  if (checkArgCount(TheCall, 1))
+    return ExprError();
+
+  ExprResult ArgRes = UsualUnaryConversions(TheCall->getArg(0));
+  if (ArgRes.isInvalid())
+    return ExprError();
+
+  const Expr *Arg = ArgRes.get()->IgnoreParenImpCasts();
+  if (!isa<PointerType>(Arg->getType()) && !Arg->getType()->isArrayType())
+    return Diag(Arg->getBeginLoc(),
+                diag::err_builtin_counted_by_ref_must_be_flex_array_member)
+           << Arg->getSourceRange();
+
+  if (Arg->HasSideEffects(Context))
+    Diag(Arg->getBeginLoc(), diag::warn_builtin_counted_by_ref_has_side_effects)
+        << Arg->getSourceRange();
+
+  // See if we have something like '&ptr->fam[0]`.
+  if (auto *UO = dyn_cast<UnaryOperator>(Arg);
+      UO && UO->getOpcode() == UO_AddrOf) {
+    Arg = UO->getSubExpr()->IgnoreParenImpCasts();
+
+    if (auto *ASE = dyn_cast<ArraySubscriptExpr>(Arg))
+      Arg = ASE->getBase()->IgnoreParenImpCasts();
+  }
+
+  if (const MemberExpr *ME = dyn_cast_if_present<MemberExpr>(Arg)) {
+    if (!ME->isFlexibleArrayMemberLike(
+            Context, getLangOpts().getStrictFlexArraysLevel()))
+      return Diag(Arg->getBeginLoc(),
+                  diag::err_builtin_counted_by_ref_must_be_flex_array_member)
+             << Arg->getSourceRange();
+
+    if (ME->getMemberDecl()->getType()->isCountAttributedType()) {
```

Comment:
`CountAttributedType` also includes `sized_by`, `counted_by_or_null`, and `sized_by_or_null`.  So you may want to do this:
```
if (auto *CATy = ME->getMemberDecl()->getType()->getAs<CountAttributedType>()) {
  if (CATy->isOrNull() || CATy->isCountInBytes()) {
    // Diag error: the builtin only supports counted_by! 
  }
```

Reply:
I'll improve the check, however it's not an error for the argument to be a flexible array member without the `counted_by` attribute. I'll just let it fall through to returning NULL.